### ISSUE TITLE
Add non-synthetic FS tests

### ISSUE THREAD

**Issue by l0kod** [landlock-lsm/linux#21]
We currently test a set of synthetic filesystems (e.g. tmpfs, proc, sysfs) thanks to the [`layout3_fs` test variants](https://github.com/landlock-lsm/linux/blob/v6.7/tools/testing/selftests/landlock/fs_test.c#L4557-L4615).

It would be useful to add new tests for non-synthetic filesystems (e.g. ext4, btrfs, f2fs, erofs) to make sure they are consistent with Landlock sandboxing. One of the challenges is to cleanly integrate light filesystem creations with the tests.

**Comment by abhishek23tiwari** [comment]
Hi, Can I work on this? If yes, please guide me on how to fix it. Thanks

**Comment by l0kod** [comment]
Hi @abhishek23tiwari! Yes you can start working this task.

This is not a bug but a test improvement. The first step would be to create minimal filesystems with `mkfs.*`, then mount them, and after that extend the `layout3_fs` test variants. All that needs to happen in the `fs_test.c`. Does it look good?

**Comment by abhishek23tiwari** [comment]
Hi, I am one of the Outreachy Applicants and I am not able to access tut - http://kernelnewbies.org/Outreachyfirstpatch. Just found that I need to complete this before working on this task. I mailed them but not sure about it because using git to send email for the first time. We can delete this message later to maintain the consistency of this task but I need little help. 
Thanks

**Comment by Chand-ra** [comment]
Hey @l0kod, I was trying to work on this issue and here's the steps I took:

1.  Use `mkfs.ntfs /dev/sdb2` to create a NTFS-filesystem.
2. Use `mount /dev/sdb2 /test` to mount the new filesystem to `/test`.
3. Add the following code to `fs_test.c`:
```
FIXTURE_VARIANT_ADD(layout3_fs, ntfs) {
	.mnt = {
		.type = "ntfs",
	},
	.file_path = TMP_DIR "/test",
};
```
4. Build the resulting system with `make`.

But the last step executes without any output. I also wanted to test the changes with landlock-test-tools but am getting the following error: `sort: cannot read: tools/testing/selftests/landlock/config: No such file or directory`.
Any pointers on the next steps I should take?

**Comment by demonkillerr** [comment]
Hey @l0kod. IRL, the fs needs to be created and mounted against a said volume. For example

`mkfs.ext4 /dev/sdaX` 
and 
`mount /dev/sdaX /mnt` for example

Can you please point me on how to create the minimal fs with respect to layout3_fs?

**Comment by l0kod** [comment]
> Hey @l0kod.

Hi!

> IRL, the fs needs to be created and mounted against a said volume. For example
> 
> `mkfs.ext4 /dev/sdaX` and `mount /dev/sdaX /mnt` for example
> 
> Can you please point me on how to create the minimal fs with respect to layout3_fs?

We can extend `FIXTURE(layout3_fs)` with some fields to identify the command needed to create a file system, and then in `FIXTURE_SETUP(layout3_fs)` create this file system accordingly. However, instead of using `/dev/sdaX` we should use a loop device:
```
fallocate --length 4M test-ext4.img
mkfs.ext4 -q test-ext4.img mnt
```

**Comment by MOLeCULE-Soft** [comment]
Greetings @l0kod, I am an Outreachy applicant and have had 3 patches accepted already. I would like to work on this.
Thanks

**Comment by l0kod** [comment]
Hi @MOLeCULE-Soft,
You can work on this. Because several people are interested by this task, everyone should first start with a specific filesystem. We'll see how it goes when patches will be posted on the mailing list.

@abhishek23tiwari @demonkillerr, what is your status? Are you still interested?

**Comment by demonkillerr** [comment]
Hi @l0kod. I am not part of the Outreachy program, so please feel free to assign it to others. I am a distro maintainer looking at some kernel code 😅. 

I am still interested to help regardless.

**Comment by Tundroid** [comment]
> Hi @MOLeCULE-Soft, 
Hello @l0kod, sorry I used the wrong account to send the first message, this is my active GitHub account now. I only realised this after your reply.
Thanks.

**Comment by l0kod** [comment]
> Hi @l0kod. I am not part of the Outreachy program, so please feel free to assign it to others. I am a distro maintainer looking at some kernel code 😅.

OK
 
> I am still interested to help regardless.

Good! You might be interested by other [test tasks](https://github.com/landlock-lsm/linux/issues?q=is%3Aissue+is%3Aopen+label%3Atest). They may be a bit more complex though.

**Comment by Tundroid** [comment]
Hello @l0kod 

I am having issues creating a loop device for ext4.

`system("losetup /dev/loop0 test-ext4.img");`

Using system above, I get the following error

`losetup: /dev/loop0: failed to set up loop device: No such device or address`

And when I use 

`open("/dev/loop0", O_RDWR);`

I get an error still.
Please, can you direct me on what I am missing.

Thanks in advance

**Comment by saashaspace** [comment]
Hello Sir @l0kod,

If we are creating a file system with mkfs.*, do we need to check for any condition where there might be a failure, (say `PID = -1`)

**Comment by abhishek23tiwari** [comment]
> Hi @MOLeCULE-Soft, You can work on this. Because several people are interested by this task, everyone should first start with a specific filesystem. We'll see how it goes when patches will be posted on the mailing list.
> 
> @abhishek23tiwari @demonkillerr, what is your status? Are you still interested?

Hi, because of some reasons, I am not able to continue with the program. And I think you already assigned someone so, I have no issue. :) Happy Coding.

**Comment by l0kod** [comment]
> when we are creating a file system with mkfs.*, do we need to check for any condition where there might be a failure, (say `PID = -1`)

Yes, all steps required to create a file system must be checked with asserts.

**Comment by l0kod** [comment]
I guess @Tundroid is not working on this task anymore.

**Comment by micromaomao** [comment]
Hi @l0kod 
On the topic of filesystems, one interesting (bug?) thing I noticed is that 9pfs seems to break landlock rules, and I realized it was because 9pfs seems to for some reason not respect the inode held by landlock, and will create new ones on attempt to access a dir which isn't opened by anyone else. For example:

```
root@1af6e5dc0b82 ~# env LL_FS_RO=/etc:/usr:/bin:/lib LL_FS_RW= /sandboxer bash
Executing the sandboxed command...
bash: /etc/bash.bashrc: Permission denied
bash-5.2# ls -la /etc
ls: cannot open directory '/etc': Permission denied
bash-5.2# 
exit
root@1af6e5dc0b82 ~ [2]# cd /etc # keep /etc open, somehow
root@1af6e5dc0b82 /etc# env LL_FS_RO=/etc:/usr:/bin:/lib LL_FS_RW= /sandboxer bash
Executing the sandboxed command...
root@1af6e5dc0b82:/etc# ls -la /etc | head
total 372
drwxr-xr-x 65 root root    2620 Mar  2 23:41 .
drwxrwxrwt 19 root root     520 Mar  2 23:45 ..
-rw-------  1 root root       0 Feb 24 00:00 .pwd.lock
drwxr-xr-x  3 root root      60 Mar  1 21:54 X11
-rw-r--r--  1 root root    3040 May 25  2023 adduser.conf
drwxr-xr-x  2 root root    1940 Mar  1 21:55 alternatives
drwxr-xr-x  3 root root     100 Mar  1 21:55 apparmor.d
drwxr-xr-x  8 root root     160 Feb 24 00:00 apt
-rw-r--r--  1 root root    1994 Mar 29  2024 bash.bashrc
root@1af6e5dc0b82:/etc#  
exit
root@1af6e5dc0b82 /etc# mount | grep 9p
root on / type 9p (rw,relatime,access=client,trans=virtio)
```

So I think testing 9pfs somehow could be interesting, however I'm not sure how one would actually mount a 9pfs in a selftest... (maybe this is not really doable without a qemu-based test infrastructure)

(btw, are you still updating the LSG Skillforge thing? I stumbled upon it which inspired me to look at landlock, but some of the tasks there might be stale / done?)

**Comment by l0kod** [comment]
> Hi [@l0kod](https://github.com/l0kod)

Hi!

> On the topic of filesystems, one interesting (bug?) thing I noticed is that 9pfs seems to break landlock rules, and I realized it was because 9pfs seems to for some reason not respect the inode held by landlock (I had some debug code that prints the address of the struct inode and it changes on repeated access even though landlock should be holding it), and will create new ones on attempt to access a dir which isn't opened by anyone else. For example:
> 
> ```
> root@1af6e5dc0b82 ~# env LL_FS_RO=/etc:/usr:/bin:/lib LL_FS_RW= /sandboxer bash
> Executing the sandboxed command...
> bash: /etc/bash.bashrc: Permission denied
> bash-5.2# ls -la /etc
> ls: cannot open directory '/etc': Permission denied
> bash-5.2# 
> exit
> root@1af6e5dc0b82 ~ [2]# cd /etc # keep /etc open, somehow
> root@1af6e5dc0b82 /etc# env LL_FS_RO=/etc:/usr:/bin:/lib LL_FS_RW= /sandboxer bash
> Executing the sandboxed command...
> root@1af6e5dc0b82:/etc# ls -la /etc | head
> total 372
> drwxr-xr-x 65 root root    2620 Mar  2 23:41 .
> drwxrwxrwt 19 root root     520 Mar  2 23:45 ..
> -rw-------  1 root root       0 Feb 24 00:00 .pwd.lock
> drwxr-xr-x  3 root root      60 Mar  1 21:54 X11
> -rw-r--r--  1 root root    3040 May 25  2023 adduser.conf
> drwxr-xr-x  2 root root    1940 Mar  1 21:55 alternatives
> drwxr-xr-x  3 root root     100 Mar  1 21:55 apparmor.d
> drwxr-xr-x  8 root root     160 Feb 24 00:00 apt
> -rw-r--r--  1 root root    1994 Mar 29  2024 bash.bashrc
> root@1af6e5dc0b82:/etc#  
> exit
> root@1af6e5dc0b82 /etc# mount | grep 9p
> root on / type 9p (rw,relatime,access=client,trans=virtio)
> ```

Indeed, this reminds me this other bug which is now fixed: [hostfs: Fix ephemeral inodes](http://git.kernel.org/torvalds/c/74ce793bcbde5cef0f82d6ccb3c47cb651295a9a)

We should fix this 9p bug as well. Are you interested to work on this?

> 
> So I think testing 9pfs somehow could be interesting, however I'm not sure how one would actually mount a 9pfs in a selftest... (maybe this is not really doable without a qemu-based test infrastructure)

According to the [kernel documentation](https://docs.kernel.org/filesystems/9p.html), we might be able to use [Plan 9 From User Space applications](https://github.com/9fans/plan9port).

> 
> (btw, are you still updating the LSG Skillforge thing? I stumbled upon it which inspired me to look at landlock - wanted to let you know :) (but some of the tasks there might be stale / done I think))

They are up-to-date now and we can extend them, let's get in touch. ;)

**Comment by micromaomao** [comment]
> We should fix this 9p bug as well. Are you interested to work on this?

yes this sounds interesting, I can only work on kernel stuff off my own time though so I will have a look Fri evening / next weekend (I will read the hostfs fix first)
